tools/build.py

```diff
@@ -0,0 +1,681 @@
+    """Returns the absolute path of an executable."""
+
+    exe_path = None
+    if command_or_path:
+        executable_path = shutil.which(str(command_or_path))
+        if executable_path is None:
+            raise UsageError(f"Failed to resolve executable path for '{command_or_path}'.")
+
+        exe_path = Path(executable_path)
+
+    return exe_path
+
+
+def _get_build_config_dir(build_dir: Path, config: str):
+    # build directory per configuration
+    return build_dir / config
+
+
+def _run_subprocess(args: List[str], cwd: Path = None, capture_stdout=False, shell=False, env=None,
+                    python_path: Path = None):
+
+    if isinstance(args, str):
+        raise ValueError("args should be a sequence of strings, not a string")
+
+    if env is None:
+        env = {}
+
+    my_env = os.environ.copy()
+
+    if python_path:
+        python_path = str(python_path.resolve())
+        if "PYTHONPATH" in my_env:
+            my_env["PYTHONPATH"] += os.pathsep + python_path
+        else:
+            my_env["PYTHONPATH"] = python_path
+
+    my_env.update(env)
+
+    return run(*args, cwd=cwd, capture_stdout=capture_stdout, shell=shell, env=my_env)
+
+
+def _flatten_arg_list(nested_list: List[List[str]]):
+    return [i for j in nested_list for i in j] if nested_list else []
+
+
+def _is_cross_compiling_on_apple(args):
+    if is_macOS():
+        return args.ios or args.osx_arch != platform.machine()
+
+    return False
+
+
+def _validate_unit_test_args(args):
+    if not args.onnxruntime_lib_dir:
+        raise UsageError("onnxruntime_lib_dir must be specified if enable_unit_tests is True")
+
+    ort_lib_dir = args.onnxruntime_lib_dir.resolve(strict=True)
+    if not ort_lib_dir.is_dir():
+        raise UsageError("onnxruntime_lib_dir must be a directory")
+
+    return ort_lib_dir
+
+
+def _generate_selected_ops_config(config_file: Path):
+    config_file.resolve(strict=True)
+    script = REPO_DIR / "tools" / "gen_selectedops.py"
+    _run_subprocess([sys.executable, str(script), str(config_file)])
+
+
+def _generate_build_tree(cmake_path: Path,
+                         source_dir: Path,
+                         build_dir: Path,
+                         configs: Set[str],
+                         cmake_extra_defines: List[str],
+                         args,
+                         cmake_extra_args: List[str]
+                         ):
+    log.info("Generating CMake build tree")
+
+    cmake_args = [
+        str(cmake_path),
+        str(source_dir),
+        # There are two ways of locating python C API header file. "find_package(PythonLibs 3.5 REQUIRED)"
+        # and "find_package(Python 3.5 COMPONENTS Development.Module)". The first one is deprecated and it
+        # depends on the "PYTHON_EXECUTABLE" variable. The second needs "Python_EXECUTABLE". Here we set both
+        # of them to get the best compatibility.
+        "-DPython_EXECUTABLE=" + sys.executable,
```

Comment:
do we need to define: Python_EXECUTABLE, PYTHON_EXECUTABLE, OCOS_CROSS_COMPILING?

Reply:
I'll leave Python_EXECUTABLE as we do have a `find_package(python3 ...)` usage. that means the same python binary used to run this script is guaranteed to be used for the build.